lib/portlayer/storage/volume_cache.go

```diff
@@ -35,22 +36,68 @@ type VolumeLookupCache struct {
 	vlc     map[string]Volume
 	vlcLock sync.RWMutex
 
-	// The underlying data storage implementation
-	volumeStore VolumeStorer
+	// Maps the service url of the volume store to the underlying data storage implementation
+	volumeStores map[url.URL]VolumeStorer
 }
 
-func NewVolumeLookupCache(op trace.Operation, vs VolumeStorer) (*VolumeLookupCache, error) {
+func NewVolumeLookupCache(op trace.Operation) *VolumeLookupCache {
 	v := &VolumeLookupCache{
-		vlc:         make(map[string]Volume),
-		volumeStore: vs,
+		vlc:          make(map[string]Volume),
+		volumeStores: make(map[url.URL]VolumeStorer),
 	}
 
-	return v, v.rebuildCache(op)
+	return v
 }
 
-// List the configured volume stores
-func (v *VolumeLookupCache) VolumeStoresList(op trace.Operation) (map[string]url.URL, error) {
-	return v.volumeStore.VolumeStoresList(op)
+// AddStore adds a volumestore by name.  The url returned is the service url to the volume store.
+func (v *VolumeLookupCache) AddStore(op trace.Operation, storeName string, vs VolumeStorer) (*url.URL, error) {
+	v.vlcLock.Lock()
+	defer v.vlcLock.Unlock()
+
+	// get the service url
+	u, err := util.VolumeStoreNameToURL(storeName)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := v.volumeStores[*u]; ok {
+		return nil, fmt.Errorf("volumestore (%s) already added", u.String())
+	}
+
+	v.volumeStores[*u] = vs
+	return u, v.rebuildCache(op)
+}
+
+func (v *VolumeLookupCache) volumeStore(store *url.URL) (VolumeStorer, error) {
+
+	// find the datastore
+	vs, ok := v.volumeStores[*store]
+	if !ok {
+		return nil, VolumeStoreNotFoundError{Msg: fmt.Sprintf("volume store (%s) not found", store.String())}
+	}
+
+	return vs, nil
+}
+
+// VolumeStoresList returns a list of volume store names
+func (v *VolumeLookupCache) VolumeStoresList(op trace.Operation) ([]string, error) {
+
+	v.vlcLock.RLock()
+	defer v.vlcLock.RUnlock()
+
+	stores := make([]string, len(v.volumeStores))
```

Comment:
ouch! you create an array of empty strings and then successfully just append new elements. Is it intended?
This is the make api ```make([]T, len, cap)```
I think you wanted to do something like this: ```make([]string, 0, len(v.volumeStores))```